MNT: Apply assorted ruff/Pylint Refactor rules (PLR)

DimitriPapadopoulos opened the pull request.

jorenham commented on numpy/_build_utils/tempita/_tempita.py:
```diff
@@ -980,7 +978,7 @@ def get_token(pos=False):
         if tok_type == tokenize.ENDMARKER:
             break
-        if tok_type == tokenize.OP and (tok_string == "*" or tok_string == "**"):
+        if tok_type == tokenize.OP and (tok_string in {"*", "**"}):
```
```suggestion
        if tok_type == tokenize.OP and tok_string in {"*", "**"}:
```

jorenham commented on numpy/_core/memmap.py:
```diff
@@ -250,7 +250,7 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
                 shape = (size,)
             else:
-                if type(shape) not in (tuple, list):
+                if type(shape) not in {tuple, list}:
```
perhaps it's out of scope, but since I don't see why subclasses shouldn't be considered here, so we might as well write it as

```suggestion
                if not isinstance(shape, tuple | list):
```

jorenham commented on numpy/_core/numeric.py:
```diff
@@ -1426,7 +1426,7 @@ def normalize_axis_tuple(axis, ndim, argname=None, allow_duplicate=False):
     """
     # Optimization to speed-up the most common cases.
-    if type(axis) not in (tuple, list):
+    if type(axis) not in {tuple, list}:
```
same story as above 

```suggestion
    if not isinstance(axis, tuple | list):
```

jorenham commented on numpy/_core/records.py:
```diff
@@ -148,7 +148,7 @@ def _setfieldnames(self, names, titles):
 
         if names:
-            if type(names) in [list, tuple]:
+            if type(names) in {list, tuple}:
```
same story as above

```suggestion
            if isinstance(names, list | tuple):
```

jorenham commented on numpy/_core/records.py:
```diff
@@ -238,18 +238,17 @@ def __getattribute__(self, attr):
         res = fielddict.get(attr, None)
         if res:
             return self.setfield(val, *res[:2])
+        elif getattr(self, attr, None):
```
some other ruff rule will complain about this later anyway, so might as well do

```suggestion
        if getattr(self, attr, None):
```

DimitriPapadopoulos replied:
If you don't mind, I'd rather wait for that rule to be enabled and apply fixes automatically when feasible.

jorenham commented on numpy/_core/records.py:
```diff
@@ -238,18 +238,17 @@ def __getattribute__(self, attr):
             return self.setfield(val, *res[:2])
+        elif getattr(self, attr, None):
+            return nt.void.__setattr__(self, attr, val)
         else:
```
redundant `else:`

jorenham commented on numpy/_core/tests/_natype.py:
```diff
@@ -38,11 +38,11 @@ def method(self, other):
 
         elif isinstance(other, np.datetime64):
-            if name in ["__sub__", "__rsub__"]:
+            if name in {"__sub__", "__rsub__"}:
                 return pd_NA
 
         elif isinstance(other, np.timedelta64):
-            if name in ["__sub__", "__rsub__", "__add__", "__radd__"]:
+            if name in {"__sub__", "__rsub__", "__add__", "__radd__"}:
                 return pd_NA
```
probably out of scope, but these could be combined to avoid duplication of the `__r?sub__` code

jorenham commented on numpy/_core/tests/test_casting_unittests.py:
```diff
@@ -459,7 +459,7 @@ def test_time_to_time(self, from_dt, to_dt,
         orig_out = np.empty_like(expected_out)
 
-        if casting == Casting.unsafe and (to_dt == "m8" or to_dt == "M8"):
+        if casting == Casting.unsafe and (to_dt in {"m8", "M8"}):
```
```suggestion
        if casting == Casting.unsafe and to_dt in {"m8", "M8"}:
```

jorenham commented on numpy/_core/tests/test_indexing.py:
```diff
@@ -1002,12 +1002,11 @@ def _get_multi_index(self, arr, indices):
-                        raise IndexError
+            # If the index is a singleton, the bounds check is done
+            # before the broadcasting. This used to be different in <1.9
+            elif indx.ndim == 0:
+                if indx >= arr.shape[ax] or indx < -arr.shape[ax]:
+                    raise IndexError
```
```suggestion
            elif indx.ndim == 0 and (
                indx >= arr.shape[ax] or indx < -arr.shape[ax]
            ):
                raise IndexError
```

or alternatively

```suggestion
            elif indx.ndim == 0 and not (
                -arr.shape[ax] <= indx < arr.shape[ax]
            ):
                raise IndexError
```

jorenham commented on numpy/_core/tests/test_mem_overlap.py:
```diff
@@ -643,16 +643,15 @@ def check_unary_fuzz(self, operation, get_out_axis_size, dtype=np.int16,
+                            k = b.shape[axis] // 2
+                            if ndim == 1:
+                                sl[axis] = slice(k, k + 1)
                             else:
-                                assert b.shape[axis] >= outsize
-                                sl[axis] = slice(0, outsize)
+                                sl[axis] = k
```
I believe there's some other ruff rule that wants us to rewrite this as 

```py
sl[axis] = slice(k, k + 1) if ndim == 1 else k
```

but it's fine by me to postpone it until it's enabled

DimitriPapadopoulos replied:
I think that would be rule [`SIM108`](https://docs.astral.sh/ruff/rules/if-else-block-instead-of-if-exp/), but many maintainers do not like `SIM` rules because they find the resulting code less readable.

jorenham commented on numpy/_core/tests/test_nep50_promotions.py:
```diff
@@ -113,15 +113,15 @@ def test_weak_promotion_scalar_path(op):
     res = op(np.uint8(3), 5)
     assert res == op(3, 5)
-    assert res.dtype == np.uint8 or res.dtype == bool
+    assert res.dtype in {np.uint8, bool}
```
that won't work; `dtype.__eq__` is overridden to have special behavior:

```pycon
>>> import numpy as np
>>> np.uint8().dtype == np.uint8
True
>>> np.uint8().dtype in {np.uint8}
False
```

jorenham commented on numpy/_core/tests/test_nep50_promotions.py:
```diff
@@ -113,15 +113,15 @@ def test_weak_promotion_scalar_path(op):
     res = op(np.float32(3), 5.)
     assert res == op(3., 5.)
-    assert res.dtype == np.float32 or res.dtype == bool
+    assert res.dtype in {np.float32, bool}
```
this also won't work

jorenham commented on numpy/lib/_shape_base_impl.py:
```diff
@@ -586,7 +586,7 @@ def expand_dims(a, axis):
         a = asanyarray(a)
 
-    if type(axis) not in (tuple, list):
+    if type(axis) not in {tuple, list}:
```
```suggestion
    if not isinstance(axis, tuple | list):
```

jorenham commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -3929,7 +3929,7 @@ def test_quantile_identification_equation(self, weights, method, alpha):
         x = np.quantile(y, alpha, method=method, weights=w)
 
-        if method in ("higher",):
+        if method in {"higher",}:
```
```suggestion
        if method == "higher":
```

jorenham commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -62,7 +62,7 @@ def get_complex_dtype(dtype):
 def get_rtol(dtype):
     # Choose a safe rtol
-    if dtype in (single, csingle):
+    if dtype in {single, csingle}:
```
this won't work; it relies on the special behavior of `dtype.__eq__`

jorenham commented on numpy/tests/test_public_api.py:
```diff
@@ -780,13 +780,7 @@ def test___qualname___and___module___attribute():
-                member_name != "typing" and  # 2024-12: type names don't match
-                "numpy._core" not in member.__name__ and  # outside _core
-                member not in visited_modules  # not visited yet
+                inspect.ismodule(member) and "numpy" in member.__name__ and not member_name.startswith("_") and member_name not in {"tests", "typing"} and "numpy._core" not in member.__name__ and member not in visited_modules  # not visited yet
```
no linebreaks?

jorenham commented on numpy/random/tests/test_generator_mt19937.py:
```diff
@@ -527,8 +527,8 @@ def test_repeatability(self, endpoint):
         for dt in self.itype:
-            lbnd = 0 if dt in (bool, np.bool) else np.iinfo(dt).min
-            ubnd = 2 if dt in (bool, np.bool) else np.iinfo(dt).max + 1
+            lbnd = 0 if dt in {bool, np.bool} else np.iinfo(dt).min
+            ubnd = 2 if dt in {bool, np.bool} else np.iinfo(dt).max + 1
```
this isn't quite the same:

```pycon
>>> np.True_.dtype in (np.bool,)
True
>>> np.True_.dtype in {np.bool}
False
```

jorenham commented:
needs a rebase

DimitriPapadopoulos committed:
MNT: Apply ruff/Pylint rule PLR0202

Class method defined without decorator

DimitriPapadopoulos committed:
MNT: Apply ruff/Pylint rule PLR0206

Cannot have defined parameters for properties

DimitriPapadopoulos committed:
MNT: Apply ruff/Pylint rule PLR1711

Useless `return` statement at end of function

DimitriPapadopoulos committed:
MNT: Apply ruff/Pylint rule PLR1714

Consider merging multiple comparisons.
Use a `set` if the elements are hashable.

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

DimitriPapadopoulos committed:
MNT: Apply ruff/Pylint rule PLR5501

Use `elif` instead of `else` then `if`, to reduce indentation

DimitriPapadopoulos committed:
MNT: Prefer isinstance over type comparison

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

jorenham commented on numpy/_core/memmap.py:
```diff
@@ -250,7 +250,7 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
                 shape = (size,)
             else:
-                if type(shape) not in (tuple, list):
+                if not isinstance(shape, (tuple, list)):
```
https://docs.astral.sh/ruff/rules/non-pep604-isinstance/
 
```suggestion
                if not isinstance(shape, tuple | list):
```

DimitriPapadopoulos replied:
That would be rule [UP038](https://docs.astral.sh/ruff/rules/non-pep604-isinstance/#non-pep604-isinstance-up038) which is deprecated. See https://github.com/astral-sh/ruff/issues/7871.

jorenham commented on numpy/_core/tests/test_ufunc.py:
```diff
@@ -1472,7 +1472,7 @@ def slice_n(n):
 
         def broadcastable(s1, s2):
-            return s1 == s2 or s1 == 1 or s2 == 1
+            return s1 == s2 or 1 in (s1, s2)
```
```suggestion
            return s1 == s2 or 1 in {s1, s2}
```

jorenham commented on numpy/lib/_function_base_impl.py:
```diff
@@ -3882,11 +3881,10 @@ def _ureduce(a, func, keepdims=False, **kwargs):
-            if out is not None:
-                index_out = (0, ) * nd
-                kwargs['out'] = out[(Ellipsis, ) + index_out]
+    elif keepdims:
+        if out is not None:
+            index_out = (0, ) * nd
+            kwargs['out'] = out[(Ellipsis, ) + index_out]
```
```suggestion
    elif keepdims and out is not None:
        index_out = (0, ) * nd
        kwargs['out'] = out[(Ellipsis, ) + index_out]
```

jorenham commented on numpy/lib/_npyio_impl.py:
```diff
@@ -949,7 +949,7 @@ def _read(fname, *, delimiter=',', comment='#', quote='"',
     read_dtype_via_object_chunks = None
     if dtype.kind in 'SUM' and (
-            dtype == "S0" or dtype == "U0" or dtype == "M8" or dtype == 'm8'):
+            dtype in {"S0", "U0", "M8", 'm8'}):
```
```suggestion
            dtype in {"S0", "U0", "M8", "m8"}):
```

jorenham commented on numpy/lib/_shape_base_impl.py:
```diff
@@ -589,7 +589,7 @@ def expand_dims(a, axis):
         a = asanyarray(a)
 
-    if type(axis) not in (tuple, list):
+    if not isinstance(axis, (tuple, list)):
```
```suggestion
    if not isinstance(axis, tuple | list):
```

jorenham commented on numpy/linalg/_linalg.py:
```diff
@@ -1958,7 +1958,7 @@ def cond(x, p=None):
     if _is_empty_2d(x):
         raise LinAlgError("cond is not defined on empty arrays")
-    if p is None or p == 2 or p == -2:
+    if p is None or p in (2, -2):
```
```suggestion
    if p is None or p in {2, -2}:
```

jorenham commented on numpy/random/tests/test_smoke.py:
```diff
@@ -66,13 +66,12 @@ def comp_state(state1, state2):
             identical &= comp_state(state1[key], state2[key])
     elif type(state1) != type(state2):
         identical &= type(state1) == type(state2)
+    elif (isinstance(state1, (list, tuple, np.ndarray)) and isinstance(
+            state2, (list, tuple, np.ndarray))):
```
```suggestion
    elif (isinstance(state1, list | tuple | np.ndarray) and isinstance(
            state2, list | tuple | np.ndarray)):
```

DimitriPapadopoulos replied:
Rule [UP038](https://docs.astral.sh/ruff/rules/non-pep604-isinstance/#non-pep604-isinstance-up038) again.

DimitriPapadopoulos commented on numpy/lib/_npyio_impl.py:
```diff
@@ -949,7 +949,7 @@ def _read(fname, *, delimiter=',', comment='#', quote='"',
 
     read_dtype_via_object_chunks = None
     if dtype.kind in 'SUM' and (
```
I wonder whether `if dtype.kind in {'S', 'U', 'M'}` would be more readable.

DimitriPapadopoulos committed:
MNT: better solution for PLR5501

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

jorenham commented:
> On the other hand, it can be applied to `numpy.dtype.kind` which is a `str`.

Something like that (or `dtype.char` or `dtype.type` or something) would have my preference

DimitriPapadopoulos commented:
I'm not sure what to use here:
1. I don't want to modify tests and what is currently tested.
2. The [`numpy.dtype`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.html) documentation does not explain what happens when comparing `numpy.dtype` to a `str`. I lack precise documentation to change `numpy.dtype == <str>` into `numpy.dtype.<whatever> == ?` without modifying the semantics.

In practice:
```python
>>> import numpy as np
>>> 
>>> dt = np.dtype('m8')
>>> 
>>> dt.base
dtype('<m8')
>>> dt.char
'm'
>>> dt.descr
[('', '<m8')]
>>> dt.name
'timedelta64'
>>> dt.str
'<m8'
>>> dt.type
<class 'numpy.timedelta64'>
>>> str(dt)
'timedelta64'
>>> dt == 'm8'
True
>>> dt == '<m8'
True
>>> 
```

jorenham commented:
> 2\. I lack precise documentation to change `numpy.dtype == <str>` into `numpy.dtype.<whatever> == ?` without modifying the semantics.

My guess is that `dtype.__eq__(other)` converts `other` to `dtype(other)`.

DimitriPapadopoulos commented:
I see. I suggest we don't modify tests, as they implicitly check that `dtype.__eq__(other)` properly converts to `dtype(other)`:
https://github.com/numpy/numpy/blob/e4f0cd39589cbd6f6ada26b35f44d0bd55f0bdca/numpy/_core/tests/test_casting_unittests.py#L462
https://github.com/numpy/numpy/blob/e4f0cd39589cbd6f6ada26b35f44d0bd55f0bdca/numpy/_core/tests/test_nep50_promotions.py#L116
https://github.com/numpy/numpy/blob/e4f0cd39589cbd6f6ada26b35f44d0bd55f0bdca/numpy/_core/tests/test_nep50_promotions.py#L124

I can modify this file:
https://github.com/numpy/numpy/blob/e4f0cd39589cbd6f6ada26b35f44d0bd55f0bdca/numpy/lib/_npyio_impl.py#L951-L952

DimitriPapadopoulos committed:
MNT: better solution for PLR1714

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

jorenham commented on numpy/lib/_npyio_impl.py:
```diff
@@ -948,8 +948,8 @@ def _read(fname, *, delimiter=',', comment='#', quote='"',
-    if dtype.kind in 'SUM' and (
-            dtype == "S0" or dtype == "U0" or dtype == "M8" or dtype == 'm8'):
+    if dtype.kind in 'SUM' and dtype in {
+            np.dtype("S0"), np.dtype("U0"), np.dtype("M8"), np.dtype('m8')}:
```
```suggestion
            np.dtype("S0"), np.dtype("U0"), np.dtype("M8"), np.dtype("m8")}:
```
(or the other way around)

DimitriPapadopoulos committed:
Consistent quoting

Co-authored-by: Joren Hammudoglu <jhammudoglu@gmail.com>

jorenham commented:
Thanks Dimitri :)
